firmware/src/network_model.c

```diff
@@ -0,0 +1,837 @@
+static const char DEFAULT_HOSTNAME[] = "responder";
+static const char DEFAULT_DOMAINNAME[] = "local";
+static const uint32_t NO_DEFAULT_ROUTE = 0;
+static const uint8_t MAX_NETMASK = 32;
+static const uint32_t IPV4_UNCONFIGURED = 0;
+static const uint32_t DHCP_FAILURE_RATIO = 3; // Fail 1 / n DHCP requests.
+
+static const ResponderDefinition RESPONDER_DEFINITION;
+
+/*
+ * @brief The read only state for a network Interface.
+ */
+typedef struct {
+  const char *label;
+  uint32_t id;
+  uint16_t hardware_type;
+  uint8_t hardware_address[MAC_SIZE];
+} InterfaceDefinition;
+
+/*
+ * @brief The mutable state for an interface
+ */
+typedef struct {
+  uint32_t configured_ip;
+  uint32_t current_ip;
+  uint8_t current_netmask;
+  uint8_t configured_netmask;
+  ConfigSource config_source;
+  bool current_dhcp_mode;
+  bool current_zeroconf_mode;
+  bool configured_dhcp_mode;
+  bool configured_zeroconf_mode;
+} InterfaceState;
+
+/*
+ * @brief The network model state.
+ */
+typedef struct {
+  InterfaceState *interfaces;
+  unsigned int interface_count;
+
+  uint16_t default_interface_route;
+  uint32_t default_route;
+  uint32_t nameservers[NUMBER_OF_NAMESERVERS];
+  char hostname[DNS_HOST_NAME_SIZE];
+  char domain_name[DNS_DOMAIN_NAME_SIZE];
+} NetworkModel;
+
+// Data
+static const char ETHERNET_INTERFACE_NAME[] = "eth0";
+static const char IPSEC_INTERFACE_NAME[] = "tun0";
+static const char WIFI_INTERFACE_NAME[] = "wlan0";
+
+static const InterfaceDefinition INTERFACE_DEFINITIONS[] = {
+  {
+   .label = ETHERNET_INTERFACE_NAME,
+   .id = LAN_INTERFACE_ID,
+   .hardware_type = ETHERNET_HARDWARE_TYPE,
+   // Local admin MAC address
+   .hardware_address = {0x52, 0x12, 0x34, 0x56, 0x78, 0x9a},
+  },
+  {
+   .label = IPSEC_INTERFACE_NAME,
+   .id = IPSEC_INTERFACE_ID,
+   .hardware_type = IPSEC_HARDWARE_TYPE,
+   // No h/w address for ptp links
+   .hardware_address = {0, 0, 0, 0, 0, 0},
+  },
+  {
+   .label = WIFI_INTERFACE_NAME,
+   .id = WLAN_INTERFACE_ID,
+   .hardware_type = ETHERNET_HARDWARE_TYPE,
+   // Local admin MAC address
+   .hardware_address = {0x52, 0xab, 0xcd, 0xef, 0x01, 0x23},
+  },
+};
+
+static InterfaceState g_interfaces[NUMBER_OF_INTERFACES];
+
+static NetworkModel g_network_model;
+
+// Helper functions
+// ----------------------------------------------------------------------------
+
+static int LookupIndex(unsigned int id) {
+  unsigned int i = 0;
+  for (; i < g_network_model.interface_count; i++) {
+    if (INTERFACE_DEFINITIONS[i].id == id) {
+      return i;
+    }
+  }
+  return -1;
+}
+
+/*
+ * @brief Check a request that contains an interface id.
+ * @returns -1 if the id was found and index was set, otherwise the size of the
+ * RDM response that was created.
+ */
+static int CheckInterfaceRequest(const RDMHeader *header,
+                                 const uint8_t *param_data,
+                                 unsigned int *index) {
+  if (header->param_data_length != sizeof(uint32_t)) {
```

Comment:
Then it allows too much.  This function is only used for GETs.

As an aside, I have a better way of doing this which is coming in the PR after this one.